### PR TITLE
i-bem: store locals on reapply

### PR DESCRIPTION
**Pull request by indutny**
Fix: https://github.com/bem/bem-xjst/issues/97

### CODE REVIEW

**Comment by arikon** [comment]
@veged @tadatuta Could you make review plz?

**Comment by arikon** [comment]
@mishaberezin This should be ported to `islands` to be fixed there

**Comment by veged** [comment]
@indutny LGTM about standard fields, but what about custom fields in context?

**Comment by tadatuta** [comment]
Added custom field to test, everything works fine. Let's merge!

**Comment by arikon** [comment]
@tadatuta I think, it should be better to squash commits

**Comment by tadatuta** [comment]
@arikon squashed, rebased

**Comment by tadatuta** [comment]
@mishaberezin have you ported it to Islands?

**Comment by mishaberezin** [comment]
Nope. Going to port on next week.

File: common.blocks/i-bem/i-bem.bemhtml [hunk 1]
```diff
@@ -250,11 +250,41 @@ BEMContext.prototype.generateId = function generateId() {
...
     var ctx = new BEMContext(context || this, oldApply);
+
     ctx.apply();
+
     return ctx._str;
 };
 
-BEMContext.prototype.reapply = BEMContext.apply;
+BEMContext.prototype.reapply = function reapply() {
+    var mode = typeof $$mode !== 'undefined' && $$mode,
```
Comment (qfox) [inline common.blocks/i-bem/i-bem.bemhtml:260]: @indutny @tadatuta Why globals? What if I call reapply twice()? Shouldn't it be a some kind of states stack?

Comment (indutny) [inline common.blocks/i-bem/i-bem.bemhtml:260]: @zxqfox this was introduced for 20% speed boost. `BEMContext` is way too polymorphic, and access these properties was very slow.

Comment (qfox) [inline common.blocks/i-bem/i-bem.bemhtml:260]: We can define them all (`$$*`) as variables in the parent scope (inside `oninit`). I don't see why they should be globals and how it (making them globals) gives 20% speed boost ;-(

Comment (indutny) [inline common.blocks/i-bem/i-bem.bemhtml:260]: I forgot to say, they are not globals. They are in parent scope of `oninit()`, which is non-global.

Comment (qfox) [inline common.blocks/i-bem/i-bem.bemhtml:260]: Ah, thanks for noticing. Now it's much clearer.

**Pull request merged**